File: exercism/python/run-length-encoding/run_length.py
import re

def encode(message):
  chunks = re.findall("((.)\\2*)", message)
  message = ""
  for chunk, _ in chunks:
    message += (str(len(chunk)) if len(chunk) > 1 else "") + chunk[0]
  return message

def decode(code):
  chunks = re.findall("[0-9]+|.", code)
  message = ""
  while len(chunks) > 0:
    try:
      count = int(chunks[0])
      chunks.pop(0)
    except:
      count = 1
    message += chunks.pop(0) * count
  return message

File: exercism/haskell/run-length-encoding/src/RunLength.hs
module RunLength (decode, encode) where

import Data.List.Split (keepDelimsR, split, whenElt, dropFinalBlank)
import Data.List (group)
import Data.Char (isDigit)


encode :: String -> String
encode =
  concatMap compact . group
  where
    compact l =
      lengthString (length l) ++ [head l]
    lengthString n =
      if n > 1 then show n else ""


decode :: String -> String
decode =
  concatMap expand . splitCodes
  where
    splitCodes =
      split . dropFinalBlank . keepDelimsR . whenElt $ not . isDigit
    expand code =
      replicate (count code) (last code)
    count code =
      max 1 (read $ '0' : init code)
